osu.Game/Database/LazerImportManager.cs

```diff
@@ -0,0 +1,405 @@
+            };
+            notifications.Post(filesNotification);
+
+            await copyFiles(filesNotification, sourcePath).ConfigureAwait(false);
+
+            filesNotification.CompletionText = "Files copied!";
+            filesNotification.State = ProgressNotificationState.Completed;
+
+            if (filesNotification.State == ProgressNotificationState.Cancelled) return;
+
+            realmAccess.Write(destRealm =>
+            {
+                var sourceConfig = new RealmConfiguration(Path.Combine(sourcePath, "client.realm"))
+                {
+                    IsReadOnly = true,
+                    SchemaVersion = RealmAccess.schema_version
+                };
+
+                using (var sourceRealm = Realm.GetInstance(sourceConfig))
+                {
+                    importRealmFiles(sourceRealm, destRealm);
+
+                    importBeatmaps(sourceRealm, destRealm);
+
+                    importScores(sourceRealm, destRealm);
+
+                    importSkins(sourceRealm, destRealm);
+
+                    importCollections(sourceRealm, destRealm);
+                }
+            });
+        }
+
+        private async Task copyFiles(ProgressNotification notification, string sourcePath)
+        {
+            string sourceFilesPath = Path.Combine(sourcePath, "files");
+            var destFilesStorage = storage!.GetStorageForDirectory("files");
+
+            var sourceConfig = new RealmConfiguration(Path.Combine(sourcePath, "client.realm"))
+            {
+                IsReadOnly = true,
+                SchemaVersion = RealmAccess.schema_version
+            };
+
+            List<string> filesToCopy;
+
+            using (var realm = Realm.GetInstance(sourceConfig))
+            {
+                filesToCopy = realm.All<RealmFile>().AsEnumerable().Select(f => f.Hash).ToList();
+            }
+
+            int total = filesToCopy.Count;
+            int current = 0;
+            var uniqueHashes = new HashSet<string>(filesToCopy);
+
+            foreach (string hash in uniqueHashes)
+            {
+                if (notification.State == ProgressNotificationState.Cancelled) return;
+
+                string folder1 = hash.Substring(0, 1);
+                string folder2 = hash.Substring(0, 2);
+                string relativePath = Path.Combine(folder1, folder2, hash);
+
+                if (!destFilesStorage.Exists(relativePath))
+                {
+                    string sourceFilePath = Path.Combine(sourceFilesPath, folder1, folder2, hash);
+
+                    if (File.Exists(sourceFilePath))
+                    {
+                        using (var src = new FileStream(sourceFilePath, FileMode.Open, FileAccess.Read))
+                        using (var dst = destFilesStorage.CreateFileSafely(relativePath))
+                        {
+                            await src.CopyToAsync(dst).ConfigureAwait(false);
+                        }
```

Comment:
You probably want to use `RealmFileStore.Add` here. It has better safeties when the destination already exists, and also allows making use of hard link support when available.

Reply:
Ah this is a really weird one. Until now we've been working with `FileStream`s returned from `ArchiveReader` that seem to set the `stream.Name` to the full path (which is then [consumed here](https://github.com/ppy/osu/blob/a8a70be04ab13f0ad4c565b86ee206c19ba01d02/osu.Game/Database/RealmFileStore.cs#L66). Can you check what you're seeing in the `Name` property of the stream in this case?